2020_RTSS_Shuai_Zhao/src/parallel_provider_consumer.rs

```diff
@@ -0,0 +1,108 @@
+//! CPC model construction.
+//! Paper Information
+//! -----------------
+//! Title: DAG Scheduling and Analysis on Multiprocessor Systems: Exploitation of Parallelism and Dependency
+//! Authors: Shuai Zhao, Xiaotian Dai, Iain Bate, Alan Burns, Wanli Chang
+//! Conference: RTSS 2020
+//! -----------------
+use lib::graph_extension::*;
+use petgraph::graph::{Graph, NodeIndex};
+
+/// Algorithm 1: Step1 identifying capacity providers.
+/// capacity provider is a sub paths of the critical path
+/// See the second paragraph of IV. A. Concurrent provider and consumer model for a detailed explanation.
+#[allow(dead_code)]
+pub fn identifying_capacity_providers(mut dag: Graph<NodeData, f32>) -> Vec<Vec<NodeIndex>> {
+    let critical_path = dag.get_critical_path();
+    let mut providers = Vec::new();
+    let mut i = 0;
+    while i < critical_path.len() - 1 {
+        let mut provider = Vec::new();
+        provider.push(critical_path[i]);
+        // At least it is not an error since it is a prior node itself.
+        let mut pre_nodes = dag.get_pre_nodes(critical_path[i + 1]).unwrap();
+        while pre_nodes.len() == 1 {
+            provider.push(critical_path[i + 1]);
+            i += 1;
+            pre_nodes = dag.get_pre_nodes(critical_path[i + 1]).unwrap();
+        }
+        providers.push(provider);
+        i += 1;
+    }
+    //Sink node is processed separately because there is no subsequent node and it causes an error
+    providers.push(vec![critical_path[i]]);
+
+    providers
+}
+
+#[cfg(test)]
+
+mod tests {
+    use std::collections::HashMap;
+
+    use super::*;
+
+    fn create_sample_dag() -> Graph<NodeData, f32> {
```

Comment:
Please refactor this function so that it is clear what kind of DAG it is.

Reply:
A comment has been added to clarify the meaning.
Nodes and edges will be added to this graph later.